### PR TITLE
[AutoDiff] Disable VJP/JVP labels in `@differentiating` when marked `linear`

### PR DESCRIPTION
**Pull request by bartchr808**
When you mark a function as `linear`, you should not be able to specify a custom VJP or JVP function. Some time later this month, we are going to add a `transpose:` label in `@differentiable` that's parallel to `vjp:`. In 2-3 months, we are going to deprecate `jvp:`, `vjp`: and `transpose:` altogether to prefer `@differentiating`.

```swift
@differentiable(linear, transpose: myTransposeFunc)
func linearFunc(_ x: Float) -> Float {
    return 3 * x
}
```

This is part of an overhaul of our current AD system that will unify forward mode (will be implemented as a part of this overhaul) and reverse mode AD. 

**More info to come in the coming weeks!**

### CODE REVIEW

File: include/swift/AST/DiagnosticsParse.def [hunk 1]
```diff
@@ -1500,6 +1500,8 @@ ERROR(attr_differentiable_missing_label,PointsToFirstBadToken,
 ERROR(attr_differentiable_expected_label,none,
       "expected either 'wrt:' or a function specifier label, e.g. 'jvp:', "
       "or 'vjp:'", ())
+ERROR(attr_differentiable_no_vjp_or_jvp_when_linear,none,
+      "can't define 'vjp:' and/or 'jvp:' on functions marked as 'linear'", ())
```
Comment (dan-zheng) [inline include/swift/AST/DiagnosticsParse.def]: I might recommend preemptively mentioning `'transpose:'` (even though it isn't implemented yet).

Something along the lines of (not perfect):
```suggestion
      "cannot specify 'vjp:' or 'jvp:' for linear functions; use 'transpose:' instead", ())
```
cc @rxwei

Comment (bartchr808) [reply include/swift/AST/DiagnosticsParse.def]: Ya that's a good idea we can just add it in right away. 😄

**Review by rxwei — COMMENTED** [review]
There are two issues:

1. Handling this should be done in `visitDifferentiableAttr` in `TypeCheckAttr.cpp` as part of type checking, not as part of parsing. This is because the user input is fully compatible with the syntax, but we want to reject it for semantic reasons.

2. The description about adding a `transpose:` label is not entirely accurate. There's a short-term answer and a long-term answer to this. Some time later this month, we are going to add a `transpose:` label in `@differentiable` that's parallels to `vjp:`. In 2-3 months, we are going to deprecate `jvp:`, `vjp:` and `transpose:` altogether to prefer `@differentiating`.

**Comment by bartchr808** [comment]
> Handling this should be done in visitDifferentiableAttr in TypeCheckAttr.cpp

True that makes sense! Regarding the tests, I will move them from `differentiable_attr_parse.swift` to `differentiable_attr_type_checking.swift`. 

> The description about adding a transpose: label is not entirely accurate

Updated! 😄 Wasn't too sure about the timeline, thanks for clearing it up!

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
Missed updating an expected error test, will make the change soon and re-run tests.

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Pull request merged**